File: Java/Dagger/atm/src/main/java/ro/razvan/java/dagger/atm/CommandProcessor.java
package ro.razvan.java.dagger.atm;

import ro.razvan.java.dagger.atm.Command.Status;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.ArrayDeque;
import java.util.Deque;

@Singleton
public final class CommandProcessor {

    private final Deque<CommandRouter> commandRouterStack = new ArrayDeque<>();

    @Inject
    public CommandProcessor(CommandRouter firstCommandRouter) {
        commandRouterStack.push(firstCommandRouter);
    }

    Status process(final String input) {
        final var commandRouter = commandRouterStack.peek();

        if (commandRouter == null) {
            return Status.INVALID;
        }

        final var result = commandRouter.route(input);

        if (result.isInputCompleted()) {
            commandRouterStack.pop();
            return commandRouterStack.isEmpty()
                    ? Status.INPUT_COMPLETED
                    : Status.HANDLED;
        }

        result.nestedCommandRouter().ifPresent(commandRouterStack::push);
        return result.status();
    }
}

File: Kotlin/Dagger/atm/src/main/kotlin/ro/razvan/kotlin/dagger/atm/CommandProcessor.kt
package ro.razvan.kotlin.dagger.atm

import ro.razvan.kotlin.dagger.atm.Command.Status
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CommandProcessor @Inject constructor(firstCommandRouter: CommandRouter) {

    private val commandRouterStack: Deque<CommandRouter> = ArrayDeque<CommandRouter>()

    init {
        commandRouterStack.push(firstCommandRouter)
    }

    fun process(input: String): Status {
        val result = commandRouterStack.peek().route(input)
        if (result.status == Status.INPUT_COMPLETED) {
            commandRouterStack.pop()
            return if (commandRouterStack.isEmpty())
                Status.INPUT_COMPLETED
            else
                Status.HANDLED
        }

        result.nestedCommandRouter?.let(commandRouterStack::push)
        return result.status
    }
}